File: experiment-23-devstral/bookshop/runs/agent=hermes-local_language=typescript_prompt=neutral/rep1/book-api/tests/api.test.ts
import request from 'supertest';
import express from 'express';
import bodyParser from 'body-parser';
import { expect } from 'chai';
import {
  initializeDatabase,
  getAllBooks,
  getBookById,
  createBook,
  updateBook,
  deleteBook
} from '../src/database';

const app = express();
app.use(bodyParser.json());

app.get('/health', (req, res) => {
  res.status(200).json({ status: 'OK' });
});

app.get('/books', async (req, res) => {
  try {
    const author = req.query.author as string;
    const books = await getAllBooks(author);
    res.status(200).json(books);
  } catch (error) {
    res.status(500).json({ error: 'Internal server error' });
  }
});

app.get('/books/:id', async (req, res) => {
  try {
    const id = parseInt(req.params.id, 10);
    if (isNaN(id)) {
      return res.status(400).json({ error: 'Invalid ID' });
    }
    
    const book = await getBookById(id);
    if (!book) {
      return res.status(404).json({ error: 'Book not found' });
    }
    
    res.status(200).json(book);
  } catch (error) {
    res.status(500).json({ error: 'Internal server error' });
  }
});

app.post('/books', async (req, res) => {
  try {
    const { title, author, year, isbn } = req.body;
    
    if (!title || !author) {
      return res.status(400).json({ error: 'Title and author are required' });
    }
    
    const id = await createBook(title, author, year, isbn);
    const newBook = await getBookById(id);
    
    res.status(201).json(newBook);
  } catch (error) {
    res.status(500).json({ error: 'Internal server error' });
  }
});

app.put('/books/:id', async (req, res) => {
  try {
    const id = parseInt(req.params.id, 10);
    if (isNaN(id)) {
      return res.status(400).json({ error: 'Invalid ID' });
    }
    
    const { title, author, year, isbn } = req.body;
    
    if (!title && !author && year === undefined && !isbn) {
      return res.status(400).json({ error: 'No fields to update' });
    }
    
    const book = await getBookById(id);
    if (!book) {
      return res.status(404).json({ error: 'Book not found' });
    }
    
    await updateBook(id, title, author, year, isbn);
    const updatedBook = await getBookById(id);
    
    res.status(200).json(updatedBook);
  } catch (error) {
    res.status(500).json({ error: 'Internal server error' });
  }
});

app.delete('/books/:id', async (req, res) => {
  try {
    const id = parseInt(req.params.id, 10);
    if (isNaN(id)) {
      return res.status(400).json({ error: 'Invalid ID' });
    }
    
    const book = await getBookById(id);
    if (!book) {
      return res.status(404).json({ error: 'Book not found' });
    }
    
    await deleteBook(id);
    
    res.status(204).send();
  } catch (error) {
    res.status(500).json({ error: 'Internal server error' });
  }
});

describe('Book API', function () {
  this.timeout(5000);
  
  before(async () => {
    await initializeDatabase();
  });
  
  it('should return health check', async () => {
    const res = await request(app).get('/health');
    expect(res.status).to.equal(200);
    expect(res.body.status).to.equal('OK');
  });
  
  it('should create a book', async () => {
    const res = await request(app)
      .post('/books')
      .send({ title: 'API Test Book', author: 'API Test Author', year: 2023, isbn: '987-6543210987' });
    
    expect(res.status).to.equal(201);
    expect(res.body).to.have.property('id');
    expect(res.body.title).to.equal('API Test Book');
    expect(res.body.author).to.equal('API Test Author');
    expect(res.body.year).to.equal(2023);
    expect(res.body.isbn).to.equal('987-6543210987');
  });
  
  it('should return validation error for missing required fields', async () => {
    const res = await request(app)
      .post('/books')
      .send({ year: 2023 });
    
    expect(res.status).to.equal(400);
    expect(res.body.error).to.equal('Title and author are required');
  });
  
  it('should get all books', async () => {
    const res = await request(app).get('/books');
    expect(res.status).to.equal(200);
    expect(res.body).to.be.an('array').that.is.not.empty;
  });
  
  it('should filter books by author', async () => {
    await request(app)
      .post('/books')
      .send({ title: 'Filtered Book', author: 'Filtered Author' });
    
    const res = await request(app).get('/books').query({ author: 'API Test Author' });
    expect(res.status).to.equal(200);
    expect(res.body).to.have.lengthOf(1);
    expect(res.body[0].title).to.equal('API Test Book');
  });
  
  it('should get a book by ID', async () => {
    const res = await request(app).get('/books');
    const bookId = res.body[0].id;
    
    const bookRes = await request(app).get(`/books/${bookId}`);
    expect(bookRes.status).to.equal(200);
    expect(bookRes.body).to.have.property('id', bookId);
  });
  
  it('should return 404 for non-existent book', async () => {
    const res = await request(app).get('/books/99999');
    expect(res.status).to.equal(404);
    expect(res.body.error).to.equal('Book not found');
  });
  
  it('should update a book', async () => {
    const res = await request(app).get('/books');
    const bookId = res.body[0].id;
    
    const updateRes = await request(app)
      .put(`/books/${bookId}`)
      .send({ title: 'Updated API Book' });
    
    expect(updateRes.status).to.equal(200);
    expect(updateRes.body.title).to.equal('Updated API Book');
  });
  
  it('should return validation error for invalid ID', async () => {
    const res = await request(app).put('/books/invalid').send({ title: 'Updated Book' });
    expect(res.status).to.equal(400);
    expect(res.body.error).to.equal('Invalid ID');
  });
  
  it('should delete a book', async () => {
    const res = await request(app).get('/books');
    const bookId = res.body[0].id;
    
    const deleteRes = await request(app).delete(`/books/${bookId}`);
    expect(deleteRes.status).to.equal(204);
    
    const bookRes = await request(app).get(`/books/${bookId}`);
    expect(bookRes.status).to.equal(404);
  });
});

File: experiments/adrianco/experiment-23-devstral/bookshop/runs/agent=hermes-local_language=rust_prompt=neutral/rep1-failed/book_api/tests/api_test.rs
use actix_web::{web, App, HttpServer, HttpResponse, test, ResponseError};
use serde_json::json;
use sqlx::sqlite::SqlitePoolOptions;
use sqlx::SqlitePool;
use uuid::Uuid;
use validator::Validate;

#[derive(serde::Serialize, serde::Deserialize, Validate, Clone)]
struct Book {
    id: Option<Uuid>,
    title: String,
    author: String,
    year: Option<i32>,
    isbn: Option<String>,
}

async fn create_book(
    pool: web::Data<SqlitePool>,
    book: web::Json<Book>
) -> Result<HttpResponse, actix_web::Error> {
    if !book.validate()? {
        return Ok(HttpResponse::BadRequest().json("Invalid input"));
    }

    let query = "INSERT INTO books (id, title, author, year, isbn) VALUES ($1, $2, $3, $4, $5)";
    sqlx::query(query)
        .bind(&book.id.unwrap_or_else(Uuid::new_v4))
        .bind(&book.title)
        .bind(&book.author)
        .bind(&book.year)
        .bind(&book.isbn)
        .execute(&**pool)
        .await?;

    Ok(HttpResponse::Created().json(book))
}

async fn get_books(
    pool: web::Data<SqlitePool>,
    author: web::Query<Option<String>>
) -> Result<HttpResponse, actix_web::Error> {
    let mut query = String::from("SELECT * FROM books");
    if let Some(author_filter) = author.into_inner() {
        query.push_str(" WHERE author LIKE ?");
        let rows = sqlx::query_as::<_, Book>(&query)
            .bind(format!("%{}%", author_filter))
            .fetch_all(&**pool)
            .await?;
        Ok(HttpResponse::Ok().json(rows))
    } else {
        let rows = sqlx::query_as::<_, Book>(&query)
            .fetch_all(&**pool)
            .await?;
        Ok(HttpResponse::Ok().json(rows))
    }
}

async fn get_book(
    pool: web::Data<SqlitePool>,
    book_id: web::Path<Uuid>
) -> Result<HttpResponse, actix_web::Error> {
    let row = sqlx::query_as::<_, Book>("SELECT * FROM books WHERE id = ?")
        .bind(book_id.into_inner())
        .fetch_one(&**pool)
        .await?;
    Ok(HttpResponse::Ok().json(row))
}

async fn update_book(
    pool: web::Data<SqlitePool>,
    book_id: web::Path<Uuid>,
    book: web::Json<Book>
) -> Result<HttpResponse, actix_web::Error> {
    if !book.validate()? {
        return Ok(HttpResponse::BadRequest().json("Invalid input"));
    }

    let query = "UPDATE books SET title = $1, author = $2, year = $3, isbn = $4 WHERE id = $5";
    sqlx::query(query)
        .bind(&book.title)
        .bind(&book.author)
        .bind(&book.year)
        .bind(&book.isbn)
        .bind(book_id.into_inner())
        .execute(&**pool)
        .await?;

    Ok(HttpResponse::Ok().json(book))
}

async fn delete_book(
    pool: web::Data<SqlitePool>,
    book_id: web::Path<Uuid>
) -> Result<HttpResponse, actix_web::Error> {
    sqlx::query("DELETE FROM books WHERE id = ?")
        .bind(book_id.into_inner())
        .execute(&**pool)
        .await?;

    Ok(HttpResponse::Ok().json({"result": "Book deleted"}))
}

async fn health_check() -> HttpResponse {
    HttpResponse::Ok().json({"status": "healthy"})
}

#[actix_rt::test]
async fn test_api() {
    // Initialize database
    let pool = SqlitePoolOptions::new()
        .max_connections(5)
        .connect("sqlite::memory:")
        .await
        .expect("Failed to create pool.");

    // Create tables
    sqlx::query(
        "CREATE TABLE IF NOT EXISTS books (
            id TEXT PRIMARY KEY,
            title TEXT NOT NULL,
            author TEXT NOT NULL,
            year INTEGER,
            isbn TEXT
        )"
    )
    .execute(&pool)
    .await
    .expect("Failed to create table");

    // Create test server
    let server = test::init_service(
        App::new()
            .app_data(web::Data::new(pool.clone()))
            .route("/health", web::get().to(health_check))
            .route("/books", web::post().to(create_book))
            .route("/books", web::get().to(get_books))
            .route("/books/{id}", web::get().to(get_book))
            .route("/books/{id}", web::put().to(update_book))
            .route("/books/{id}", web::delete().to(delete_book))
    ).await;

    // Test health check
    let req = test::TestRequest::get().uri("/health").to_request();
    let resp = test::call_service(&server, req).await;
    assert_eq!(resp.status(), 200);

    // Test create book
    let book = Book {
        id: None,
        title: "Test Book".to_string(),
        author: "Test Author".to_string(),
        year: Some(2023),
        isbn: Some("1234567890".to_string()),
    };
    let req = test::TestRequest::post()
        .uri("/books")
        .set_json(&book)
        .to_request();
    let resp = test::call_service(&server, req).await;
    assert_eq!(resp.status(), 201);
    let body = test::read_body(resp).await;
    let created_book: Book = serde_json::from_slice(&body).unwrap();
    assert!(!created_book.id.is_none());

    // Test get book by id
    let req = test::TestRequest::get()
        .uri(&format!("/books/{}", created_book.id.unwrap()))
        .to_request();
    let resp = test::call_service(&server, req).await;
    assert_eq!(resp.status(), 200);

    // Test get all books
    let req = test::TestRequest::get().uri("/books").to_request();
    let resp = test::call_service(&server, req).await;
    assert_eq!(resp.status(), 200);

    // Test update book
    let updated_book = Book {
        id: created_book.id,
        title: "Updated Book".to_string(),
        author: "Updated Author".to_string(),
        year: Some(2024),
        isbn: Some("0987654321".to_string()),
    };
    let req = test::TestRequest::put()
        .uri(&format!("/books/{}", created_book.id.unwrap()))
        .set_json(&updated_book)
        .to_request();
    let resp = test::call_service(&server, req).await;
    assert_eq!(resp.status(), 200);

    // Test delete book
    let req = test::TestRequest::delete()
        .uri(&format!("/books/{}", created_book.id.unwrap()))
        .to_request();
    let resp = test::call_service(&server, req).await;
    assert_eq!(resp.status(), 200);
}
